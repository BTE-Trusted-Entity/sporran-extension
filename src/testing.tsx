export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
