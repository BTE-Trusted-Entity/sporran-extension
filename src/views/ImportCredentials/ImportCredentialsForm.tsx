import { ChangeEvent, DragEvent, useCallback } from 'react';
import { browser } from 'webextension-polyfill-ts';

import * as styles from './ImportCredentials.module.css';

import { Stats } from '../../components/Stats/Stats';

interface Props {
  handleFiles: (files: FileList) => void;
}

export function ImportCredentialsForm({ handleFiles }: Props): JSX.Element {
  const t = browser.i18n.getMessage;

  const handleChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const input = event.target;
      if (input.files) {
        handleFiles(input.files);
      }
    },
    [handleFiles],
  );

  const handleDrop = useCallback(
    (event: DragEvent) => {
      event.preventDefault();
      event.stopPropagation();

      handleFiles(event.dataTransfer.files);
    },
    [handleFiles],
  );

  const handleDrag = useCallback((event: DragEvent) => {
    event.preventDefault();
    event.stopPropagation();
  }, []);

  return (
    <section
      className={styles.container}
      onDragEnter={handleDrag}
      onDragOver={handleDrag}
      onDrop={handleDrop}
    >
      <h1 className={styles.heading}>{t('view_ImportCredentials_title')}</h1>
      <p className={styles.subline}>{t('view_ImportCredentials_subline')}</p>

      <div className={styles.workspace}>
        <form className={styles.form}>
          <p className={styles.drop}>{t('view_ImportCredentials_drop')}</p>

          <p>
            <label className={styles.fileLabel}>
              {t('view_ImportCredentials_cta')}
              <input
                type="file"
                className={styles.file}
                onChange={handleChange}
                name="credentials"
                multiple
                accept=".json,text/json"
              />
            </label>
          </p>

          <button
            className={styles.cancel}
            onClick={window.close}
            type="button"
          >
            {t('common_action_cancel')}
          </button>
        </form>
      </div>

      <Stats />
    </section>
  );
}
