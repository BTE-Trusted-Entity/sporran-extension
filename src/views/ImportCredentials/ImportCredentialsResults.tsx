import { MouseEvent, useCallback, useState } from 'react';
import { Modal } from 'react-dialog-polyfill';
import { browser } from 'webextension-polyfill-ts';
import { without } from 'lodash-es';

import * as styles from './ImportCredentials.module.css';

import { Identity } from '../../utilities/identities/types';
import { IdentityLine } from '../../components/IdentityLine/IdentityLine';
import { Stats } from '../../components/Stats/Stats';

import { FailedImport, Import, SuccessfulImport } from './types';

interface Props {
  grouped: {
    identity: Identity;
    imports: SuccessfulImport[];
  }[];
  pending: Import[];
  failedImports: FailedImport[];
  setFailedImports: (values: FailedImport[]) => void;
  handleMoreClick: (event: MouseEvent) => void;
}

export function ImportCredentialsResults({
  grouped,
  pending,
  failedImports,
  setFailedImports,
  handleMoreClick,
}: Props): JSX.Element {
  const t = browser.i18n.getMessage;

  const [modalText, setModalText] = useState<string>();
  const onModalClose = useCallback(() => setModalText(undefined), []);

  const errors: Record<
    FailedImport['error'],
    { label: string; message: string }
  > = {
    orphaned: {
      label: t('view_ImportCredentials_unknownIdentityLabel'),
      message: t('view_ImportCredentials_unknownIdentityInfo'),
    },
    invalid: {
      label: t('view_ImportCredentials_wrongFormatLabel'),
      message: t('view_ImportCredentials_wrongFormatInfo'),
    },
    unexpected: {
      label: t('view_ImportCredentials_unexpectedErrorLabel'),
      message: t('view_ImportCredentials_unexpectedErrorInfo'),
    },
  };

  return (
    <section
      className={styles.container}
      onDragEnter={handleMoreClick}
      onDragOver={handleMoreClick}
      onDrop={handleMoreClick}
    >
      <h1 className={styles.heading}>{t('view_ImportCredentials_title')}</h1>
      <p className={styles.subline}>{t('view_ImportCredentials_subline')}</p>

      <div className={styles.workspace}>
        {pending.length > 0 && (
          <ul className={styles.imports}>
            {pending.map((fileImport, index) => (
              <li key={`${index}${fileImport}`} className={styles.fileImport}>
                <h4 className={styles.fileName} title={fileImport.fileName}>
                  {fileImport.fileName}
                </h4>
                <p className={styles.fileResult}>
                  {t('view_ImportCredentials_pendingImport')}
                </p>
              </li>
            ))}
          </ul>
        )}

        {failedImports.length > 0 && (
          <ul className={styles.imports}>
            {failedImports.map((fileImport, index) => {
              const error = errors[fileImport.error] || errors.unexpected;
              return (
                <li
                  key={`${index}${fileImport}`}
                  className={styles.failedImport}
                >
                  <h4 className={styles.fileName} title={fileImport.fileName}>
                    {fileImport.fileName}
                  </h4>
                  <button
                    className={styles.fileHide}
                    type="button"
                    aria-label={t('view_ImportCredentials_hideError')}
                    title={t('view_ImportCredentials_hideError')}
                    onClick={() =>
                      setFailedImports(without(failedImports, fileImport))
                    }
                  />
                  <button
                    className={styles.fileResultFail}
                    type="button"
                    onClick={() => setModalText(error.message)}
                  >
                    {error.label}
                  </button>
                </li>
              );
            })}
          </ul>
        )}

        {modalText && (
          <Modal open className={styles.overlay}>
            <p className={styles.overlayText}>{modalText}</p>
            <button
              type="button"
              className={styles.overlayClose}
              onClick={onModalClose}
            >
              {t('common_action_close')}
            </button>
          </Modal>
        )}

        {grouped && grouped.length > 0 && (
          <ul className={styles.imports}>
            {grouped.map(({ identity, imports }) => (
              <li key={identity.address}>
                <IdentityLine identity={identity} />

                <ul className={styles.imports}>
                  {imports.map((fileImport, index) => (
                    <li
                      key={`${index}${fileImport}`}
                      className={styles.successfulImport}
                    >
                      <h4
                        className={styles.successfulFileName}
                        title={fileImport.fileName}
                      >
                        {fileImport.fileName}
                      </h4>
                      <p className={styles.fileResult}>
                        {t('view_ImportCredentials_successfulImport')}
                      </p>
                    </li>
                  ))}
                </ul>
              </li>
            ))}
          </ul>
        )}

        <button className={styles.more} onClick={handleMoreClick} type="button">
          {t('view_ImportCredentials_more')}
        </button>

        <button className={styles.home} onClick={window.close} type="button">
          {t('common_action_close')}
        </button>
      </div>

      <Stats />
    </section>
  );
}
