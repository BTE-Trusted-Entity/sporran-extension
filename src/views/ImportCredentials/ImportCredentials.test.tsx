import { identitiesMock, render } from '../../testing/testing';

import { ImportCredentialsForm } from './ImportCredentialsForm';
import { ImportCredentialsResults } from './ImportCredentialsResults';

describe('ImportCredentials', () => {
  describe('ImportCredentialsForm', () => {
    it('should match the snapshot', async () => {
      const { container } = render(
        <ImportCredentialsForm handleFiles={jest.fn()} />,
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('ImportCredentialsResults', () => {
    it('should match the snapshot', async () => {
      const { container } = render(
        <ImportCredentialsResults
          pending={[{ fileName: 'Email-1.json' }]}
          grouped={[
            {
              identity:
                identitiesMock[
                  '4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1'
                ],
              imports: [
                {
                  fileName: 'Twitter-1.json',
                  identityAddress:
                    '4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1',
                },
                {
                  fileName:
                    'Twitter-2-very-long-name-that-overflows-the-space.json',
                  identityAddress:
                    '4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1',
                },
              ],
            },
          ]}
          failedImports={[
            { fileName: 'Dotsama-1.json', error: 'invalid' },
            {
              fileName:
                'Dotsama-2-very-long-name-that-overflows-the-space.json',
              error: 'orphaned',
            },
            { fileName: 'Dotsama-2.json', error: 'unknown' },
          ]}
          setFailedImports={jest.fn()}
          handleMoreClick={jest.fn()}
        />,
      );
      expect(container).toMatchSnapshot();
    });
  });
});
