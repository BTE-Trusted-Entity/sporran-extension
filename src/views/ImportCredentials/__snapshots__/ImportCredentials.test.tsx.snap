// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportCredentials ImportCredentialsForm should match the snapshot 1`] = `
<div>
  <section
    class="container"
  >
    <h1
      class="heading"
    >
      Import Credentials
    </h1>
    <p
      class="subline"
    >
      To import your credentials you need your backup files
    </p>
    <div
      class="workspace"
    >
      <form
        class="form"
      >
        <p
          class="drop"
        >
          Drag’n’drop the backup files here
        </p>
        <p>
          <label
            class="fileLabel"
          >
            or Browse Files
            <input
              accept=".json,text/json"
              class="file"
              multiple=""
              name="credentials"
              type="file"
            />
          </label>
        </p>
        <button
          class="cancel"
          type="button"
        >
          Cancel
        </button>
      </form>
    </div>
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </section>
</div>
`;

exports[`ImportCredentials ImportCredentialsResults should match the snapshot 1`] = `
<div>
  <section
    class="container"
  >
    <h1
      class="heading"
    >
      Import Credentials
    </h1>
    <p
      class="subline"
    >
      To import your credentials you need your backup files
    </p>
    <div
      class="workspace"
    >
      <ul
        class="imports"
      >
        <li
          class="fileImport"
        >
          <h4
            class="fileName"
            title="Email-1.json"
          >
            Email-1.json
          </h4>
          <p
            class="fileResult"
          >
            Processing…
          </p>
        </li>
      </ul>
      <ul
        class="imports"
      >
        <li
          class="failedImport"
        >
          <h4
            class="fileName"
            title="Dotsama-1.json"
          >
            Dotsama-1.json
          </h4>
          <button
            aria-label="Hide this message"
            class="fileHide"
            title="Hide this message"
            type="button"
          />
          <button
            class="fileResultFail"
            type="button"
          >
            Wrong file format
          </button>
        </li>
        <li
          class="failedImport"
        >
          <h4
            class="fileName"
            title="Dotsama-2-very-long-name-that-overflows-the-space.json"
          >
            Dotsama-2-very-long-name-that-overflows-the-space.json
          </h4>
          <button
            aria-label="Hide this message"
            class="fileHide"
            title="Hide this message"
            type="button"
          />
          <button
            class="fileResultFail"
            type="button"
          >
            Identity missing
          </button>
        </li>
        <li
          class="failedImport"
        >
          <h4
            class="fileName"
            title="Dotsama-2.json"
          >
            Dotsama-2.json
          </h4>
          <button
            aria-label="Hide this message"
            class="fileHide"
            title="Hide this message"
            type="button"
          />
          <button
            class="fileResultFail"
            type="button"
          >
            Unexpected error
          </button>
        </li>
      </ul>
      <ul
        class="imports"
      >
        <li>
          <section
            class="component"
          >
            <section
              class="identity"
            >
              <div
                class="avatar"
              >
                Identicon
              </div>
              <p
                class="name"
              >
                Light DID Identity
              </p>
            </section>
            <span
              class="amount"
            >
              1.2340
               
              <span
                aria-label="KILT Coin"
                class="coin"
              />
            </span>
          </section>
          <ul
            class="imports"
          >
            <li
              class="successfulImport"
            >
              <h4
                class="successfulFileName"
                title="Twitter-1.json"
              >
                Twitter-1.json
              </h4>
              <p
                class="fileResult"
              >
                Done
              </p>
            </li>
            <li
              class="successfulImport"
            >
              <h4
                class="successfulFileName"
                title="Twitter-2-very-long-name-that-overflows-the-space.json"
              >
                Twitter-2-very-long-name-that-overflows-the-space.json
              </h4>
              <p
                class="fileResult"
              >
                Done
              </p>
            </li>
          </ul>
        </li>
      </ul>
      <button
        class="more"
        type="button"
      >
        Import More Credentials
      </button>
      <button
        class="home"
        type="button"
      >
        Close
      </button>
    </div>
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </section>
</div>
`;
